### PR TITLE
Add track command

### PR DESCRIPTION
**Pull request by williamjamir**
Add track command for the rename script, with this command the user can compare the difference between two branches and generate a list of the imports that was changed.

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -1,3 +1,5 @@
+skip_branch_with_pr: true
```
Comment (nicoddemus) [inline appveyor.yml:1]: Hmm this is interesting. Does `travis` has a similar option?

File: appveyor.yml [hunk 2]
```diff
@@ -9,6 +11,12 @@ environment:
     - PYTHON: "C:\\Python36"
       TOX_ENV: "py36"
 
+  global:
+    LOGNAME: "LOGNAME"
```
Comment (nicoddemus) [inline appveyor.yml]: This seems like a placeholder that was left in there

File: module_renamer/cli.py [hunk 3]
```diff
@@ -1,16 +1,34 @@
...
 
-@click.command()
-def main(args=None):
-    """Console script for module_renamer."""
-    click.echo("Replace this message by putting your code into "
-               "module_renamer.cli.main")
-    return 0
+@main.command()
+@click.argument('project_path', type=click.Path(exists=True))
+@click.option('--origin_branch', default='master',
```
Comment (nicoddemus) [inline module_renamer/cli.py]: Options usually are separated by dashes: `--origin-branch`, same for `--work-branch`.

Also, a more intuitive command-line IMO would be:

```
module-renamer --output-file=foo.py   # current branch (with modifications) compared to "master"

module-renamer --branch=my-branch --compare-with=master --output-file=foo.py   # "my-branch" with modifications compared with "master"
```

File: module_renamer/commands/track_modifications.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+Import = namedtuple("Import", ["module", "name"])
+
+
+@contextmanager
+def branch_checkout(repo, branch_name):
+    # type: (git.Repo, str) -> None
+    """
+    Change the current git branch for the branch informed on branch_name
+
+    :param repo: A git.Repo object that represents the git repository from the project
```
[the remark this answers is not in the document]
Comment (nicoddemus) [reply module_renamer/commands/track_modifications.py]: As we discussed, please configure the line length to `100` 👍

File: module_renamer/commands/track_modifications.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    echo('Generating the file {0}'.format(file_name))
+
+    with open(file_name, 'w') as file:
+        file.write("imports_to_move = ")
+        file.writelines(repr(list(list_with_modified_imports)))
+
+    FormatFile(file_name, in_place=True)
+
+
+def generate_list_with_modified_imports(origin_import_list, working_import_list):
```
[the remark this answers is not in the document]
Comment (nicoddemus) [reply module_renamer/commands/track_modifications.py]: Hmm this is complaining about the length of the function: `len("generate_list_with_modified_imports") == 35`. Please see if this is customizable as well (I don't think we should add an upper limit to the name of the function, but enforcing it to be all lower-case is good)

File: module_renamer/commands/track_modifications.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+import os
+from collections import Counter, namedtuple
+from contextlib import contextmanager
+
+from click import ClickException, confirm, echo
+from git import Repo
+from tqdm import tqdm
+from yapf.yapflib.yapf_api import FormatFile
+
+CONFLICT_MSG = "\nUnfortunately, you moved two objects with the same name on " \
```
Comment (nicoddemus) [inline module_renamer/commands/track_modifications.py]: I usually take advantage of the fact that there are implicit line continuations inside `()` so I write this:

```python
CONFLICT_MSG = (
    "\n"
    "Unfortunately, you moved two objects with the same name on different paths.\n"
    "This situation could be catastrophic while running the script for rename the imports.\n"
    "These are the imports that have conflict:\n"
    " -> {} \n"
)
```

File: module_renamer/commands/track_modifications.py (hunk 4 above)
Comment (nicoddemus) [inline module_renamer/commands/track_modifications.py]: You can add the type to the `:param:` directive directly:

```
:param git.Repo repo: the git repository of the project
```

File: module_renamer/commands/track_modifications.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    :type repo: git.Repo
+
+    :param branch_name: Name of the branch that Git should move intoN
+    :type branch_name: str
+    """
+    if (repo.is_dirty() or len(repo.untracked_files)) != 0:
+        raise ClickException("The repository is dirty, please clean it first ")
+
+    current_branch = repo.active_branch.name
+    getattr(repo.heads, branch_name).checkout()
```
Comment (nicoddemus) [inline module_renamer/commands/track_modifications.py]: I think this is clearer: `repo.branches[branch_name].checkout()`

File: module_renamer/commands/track_modifications.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+    try:
+        yield
+    finally:
+        getattr(repo.heads, current_branch).checkout()
+
+
+def track_modifications(**kwargs):
```
Comment (nicoddemus) [inline module_renamer/commands/track_modifications.py]: IMO "track" gives the idea that you will be "tracking" the modifications from now on (like "watching"), which is kind of misleading. How about "trace" or "analyse"?

Comment (nicoddemus) [inline module_renamer/commands/track_modifications.py]: Why this receives `kwargs` instead of explicit parameters `project_path`, `origin_branch` and `work_branch`?

File: module_renamer/commands/track_modifications.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+    else:
+        work_branch = repo.active_branch.name
+
+    if origin_branch == work_branch:
+        raise ClickException(
+            "Origin and Working branch are the same. "
+            "Please, change your activate branch to where you made you changes on the code, "
+            "or use the option --origin_branch and --work_branch  ."
+        )
+    file_counter = total_of_py_files_on_project(project_path)
```
Comment (nicoddemus) [inline module_renamer/commands/track_modifications.py]: You are traversing the tree to count the `.py` files and throwing away the results, only to later on traverse the tree again; makes more sense to just get all files at this point and reuse later.

File: module_renamer/commands/track_modifications.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+
+    with branch_checkout(repo, origin_branch):
```
Comment (nicoddemus) [inline module_renamer/commands/track_modifications.py]: Nice usage of a context manager. 👍

File: module_renamer/commands/track_modifications.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+               "the script for rename the imports. \n" \
+               "These are the imports that have conflict: \n -> {} \n"
+
+INFORMATIVE_CONFLICT_MSG = 'Do you want to generate the file without this import? ' \
+                           'Otherwise this script will be aborted.'
+Import = namedtuple("Import", ["module", "name"])
+
+
+@contextmanager
+def branch_checkout(repo, branch_name):
```
Comment (nicoddemus) [inline module_renamer/commands/track_modifications.py]: Better name would be `checkout_branch` (a verb); `branch_checkout` is more like a subject, which is a little weird for a context manager.

File: module_renamer/commands/track_modifications.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+    "list_output.py" but can be changed by passing the argument --output_file
+
+    The name of the list (inside the file) cannot be changed since it will be used later on the
+    script for renaming the project.
+    """
+    echo('Generating the file {0}'.format(file_name))
+
+    with open(file_name, 'w') as file:
+        file.write("imports_to_move = ")
+        file.writelines(repr(list(list_with_modified_imports)))
```
Comment (nicoddemus) [inline module_renamer/commands/track_modifications.py]: Consider using `pprint` instead to write a more readable file.

File: module_renamer/commands/track_modifications.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+
+    FormatFile(file_name, in_place=True)
```
Comment (nicoddemus) [inline module_renamer/commands/track_modifications.py]: Hmm why is this necessary? I mean, the file is temporary and will not be committed to the repository right?

File: module_renamer/commands/track_modifications.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+    imports_with_conflict = [class_name
+                             for class_name, number_of_occurrences in Counter(import_from).items()
+                             if number_of_occurrences > 1]
+    if len(imports_with_conflict) > 0:
+        echo(CONFLICT_MSG.format('\n -> '.join(imports_with_conflict)))
+
+        if confirm(INFORMATIVE_CONFLICT_MSG, abort=True):
+            list_with_modified_imports = [modified_import for modified_import in
+                                          list_with_modified_imports
+                                          if modified_import[0] not in imports_with_conflict
```
Comment (nicoddemus) [inline module_renamer/commands/track_modifications.py]: Each one of these `if`s is a linear search, depending on the number of imports this might impact the performance.

File: module_renamer/commands/track_modifications.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+    return list_with_modified_imports
+
+
+def get_imports(project_path, file_counter):
+    # type: (str) -> Import
+    """
+    Look for all .py files on the given project path and return the import statements found on
+    each file.
+
+    Note.: I inserted the TQDM here because was the only way that I could have an accurate
```
Comment (nicoddemus) [inline module_renamer/commands/track_modifications.py]: It seems the handling of TQDM could be moved to the calling code given that this function `yield`s each import it finds...

File: module_renamer/commands/track_modifications.py [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+def walk_on_py_files(folder):
+    """
+    Walk through each python files in a directory
+    """
+    for dir_path, _, files in os.walk(folder):
+        for filename in fnmatch.filter(files, '*.py'):
+            yield os.path.abspath(os.path.join(dir_path, filename))
+
+
+def total_of_py_files_on_project(project_path):
```
Comment (nicoddemus) [inline module_renamer/commands/track_modifications.py]: As I said previously, it seems like a waste to recursively iterate over all files and then throw the result away.

**Review by nicoddemus — CHANGES_REQUESTED** [review]

File: module_renamer/commands/track_modifications.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (williamjamir) [reply module_renamer/commands/track_modifications.py]: For now, I have been able to configure this REGEX just through the Codacy web interface. I will try to make this configuration explicit by using a file.

File: appveyor.yml (hunk 1 above)
Comment (williamjamir) [reply appveyor.yml:1]: I didn't find a "Skip branch" on Travis to avoid duplication, but I could find the following snippet that enables a "whitelist". So, Travis will just create jobs for master and Pull Requests to master.
```
branches:
  only:
    - "master"
```

[Source](https://github.com/travis-ci/travis-ci/issues/1147#issuecomment-160820262)

File: module_renamer/cli.py (hunk 3 above)
Comment (williamjamir) [reply module_renamer/cli.py]: I was not using the dashes because I was not being able to access it, but I just found that any internal " - " characters will be converted to underscore.
The next commit improves this inputs ;)

- [x] Fixed on commit d6b11ea0a7656b058ce8216ec9387fa30c228830

File: module_renamer/commands/track_modifications.py (hunk 6 above)
Comment (williamjamir) [reply module_renamer/commands/track_modifications.py]: I didn't know about this implicit line continuation, the readability of the message improved a lot =)

- [x] Fixed on commit d6b11ea0a7656b058ce8216ec9387fa30c228830

File: module_renamer/commands/track_modifications.py (hunk 8 above)
Comment (williamjamir) [reply module_renamer/commands/track_modifications.py]: I changed the analyze command to receive explicit the parameters ;)
- [x] Fixed on commit d6b11ea0a7656b058ce8216ec9387fa30c228830

File: appveyor.yml (hunk 2 above)
Comment (williamjamir) [reply appveyor.yml]: Actually, this was on purpose because of GitPython. A better approach was taken on commit https://github.com/ESSS/module-renamer/commit/0070645ed6494cb511f8b59d9032310f8eac718c . The file ` tox.ini` has comments and links explaining the issue that these variables solve.

- [x] Task: Remove global env

File: module_renamer/commands/track_modifications.py (hunk 4 above)
Comment (williamjamir) [reply module_renamer/commands/track_modifications.py]: - [x] Fixed on commit d6b11ea0a7656b058ce8216ec9387fa30c228830

File: module_renamer/commands/track_modifications.py (hunk 7 above)
Comment (williamjamir) [reply module_renamer/commands/track_modifications.py]: - [x] Fixed on commit d6b11ea0a7656b058ce8216ec9387fa30c228830

File: module_renamer/commands/track_modifications.py (hunk 8 above)
Comment (williamjamir) [reply module_renamer/commands/track_modifications.py]: Sure, the name "track" was really bad. I'll take your advice and change to analyze.
- [x] Fixed on commit d6b11ea0a7656b058ce8216ec9387fa30c228830

File: module_renamer/commands/track_modifications.py (hunk 11 above)
Comment (williamjamir) [reply module_renamer/commands/track_modifications.py]: - [x] Fixed on commit d6b11ea0a7656b058ce8216ec9387fa30c228830

File: module_renamer/commands/track_modifications.py (hunk 15 above)
Comment (williamjamir) [reply module_renamer/commands/track_modifications.py]: Sure, but the progress is based on the total number of files since I cannot know how many imports each file has beforehand.

File: module_renamer/commands/track_modifications.py (hunk 9 above)
Comment (williamjamir) [reply module_renamer/commands/track_modifications.py]: Following your suggestion, I'm getting all the files in a list and looking for its size when necessary to get the total number of files;)

- [x] Fixed on commit d6b11ea0a7656b058ce8216ec9387fa30c228830

File: module_renamer/commands/track_modifications.py (hunk 12 above)
Comment (williamjamir) [reply module_renamer/commands/track_modifications.py]: - [x] Fixed on commit d6b11ea0a7656b058ce8216ec9387fa30c228830

File: module_renamer/commands/track_modifications.py (hunk 13 above)
Comment (williamjamir) [reply module_renamer/commands/track_modifications.py]: This is the actual file that the scripts generate (to be used later on). `FormatFile` it's a function from YAPF to write a readable file.

I didn't know the existence of `pprint`, before your review so I changed FormatFile to pprint  =)
- [x] Fixed on commit d6b11ea0a7656b058ce8216ec9387fa30c228830

File: module_renamer/commands/track_modifications.py (hunk 16 above)
Comment (williamjamir) [reply module_renamer/commands/track_modifications.py]: - [x] Fixed on commit d6b11ea0a7656b058ce8216ec9387fa30c228830

**Review by nicoddemus — APPROVED** [review]
